infrastructure/src/main/java/org/corfudb/infrastructure/management/ClusterStateCollector.java

```diff
@@ -0,0 +1,134 @@
+package org.corfudb.infrastructure.management;
+
+import com.google.common.collect.ImmutableMap;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.management.ClusterStateContext.HeartbeatCounter;
+import org.corfudb.protocols.wireprotocol.ClusterState;
+import org.corfudb.protocols.wireprotocol.NodeState;
+import org.corfudb.protocols.wireprotocol.SequencerMetrics;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeConnectivity;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeConnectivity.ConnectionStatus;
+import org.corfudb.runtime.exceptions.WrongEpochException;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Collects information about cluster state.
+ * Provides list of servers in the cluster with higher epochs.
+ * Builds the {@link ClusterState} and local {@link NodeState}.
+ *
+ * The collectors navigates via list of requests represented as CompletableFuture-s
+ * and collecting information about cluster.
+ */
+@Builder
+@Slf4j
+public class ClusterStateCollector {
+    @NonNull
+    private final String localEndpoint;
+    @NonNull
+    private final Map<String, CompletableFuture<NodeState>> clusterState;
+
+    @NonNull
+    private final HeartbeatCounter heartbeatCounter;
+
+    /**
+     * Provides cluster state
+     * @param epoch current epoch
+     * @param sequencerMetrics sequencer metrics
+     * @return cluster state
+     */
+    public ClusterState collectClusterState(long epoch, SequencerMetrics sequencerMetrics) {
+        Map<String, NodeState> nodeStates = new HashMap<>();
+
+        nodeStates.put(localEndpoint, collectLocalNodeState(epoch, sequencerMetrics));
+        nodeStates.putAll(collectNodeStates());
+
+        return ClusterState.builder()
+                .localEndpoint(localEndpoint)
+                .nodes(ImmutableMap.copyOf(nodeStates))
+                .build();
+    }
+
+    /**
+     * Provides list of servers contains different epochs to this server.
+     *
+     * @return map of endpoint:epoch entries
+     */
+    public ImmutableMap<String, Long> collectWrongEpochs() {
+        Map<String, Long> wrongEpochs = new HashMap<>();
+
+        clusterState.forEach((server, state) -> {
+            try {
+                //Add a node to connectedNodes list if node responded with its NodeState
+                state.get();
+            } catch (Exception e) {
```

Comment:
What happens if you get Thowable? Unless you are catching it somewhere it will cause your periodic task to stop.

Reply:
We shouldn't catch Throwable right? If there is an error, you can't do much, you can only exit from your application

Reply:
We have been getting random InterruptedExceptions, so keep that in mind.